app/actions/local/channel.ts

```diff
@@ -139,32 +153,123 @@ export const selectAllMyChannelIds = async (serverUrl: string) => {
     return queryAllMyChannelIds(database);
 };
 
-export const markChannelAsViewed = async (serverUrl: string, channelId: string, prepareRecordsOnly = false) => {
-    const database = DatabaseManager.serverDatabases[serverUrl]?.database;
-    if (!database) {
+export const markChannelAsViewed = async (serverUrl: string, channelId: string, newLastPostAt?: number, prepareRecordsOnly = false) => {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
         return {error: `${serverUrl} database not found`};
     }
 
-    const member = await queryMyChannel(database, channelId);
+    const member = await queryMyChannel(operator.database, channelId);
     if (!member) {
         return {error: 'not a member'};
     }
 
+    const lastPostAt = newLastPostAt != null && newLastPostAt > member.lastPostAt ? newLastPostAt : member.lastPostAt;
     member.prepareUpdate((m) => {
         m.messageCount = 0;
         m.mentionsCount = 0;
         m.manuallyUnread = false;
         m.viewedAt = member.lastViewedAt;
+        m.lastViewedAt = lastPostAt + 1;
+        m.lastPostAt = lastPostAt;
+    });
+
+    try {
+        if (!prepareRecordsOnly) {
+            await operator.batchRecords([member]);
+        }
+
+        return {member};
+    } catch (error) {
+        return {error};
+    }
+};
+
+export const markChannelAsUnread = async (serverUrl: string, channelId: string, messageCount: number, mentionsCount: number, manuallyUnread: boolean, newLastPostAt?: number, lastViewed?: number, prepareRecordsOnly = false) => {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
+        return {error: `${serverUrl} database not found`};
+    }
+
+    const member = await queryMyChannel(operator.database, channelId);
+    if (!member) {
+        return {error: 'not a member'};
+    }
+
+    const lastPostAt = newLastPostAt != null && newLastPostAt > member.lastPostAt ? newLastPostAt : member.lastPostAt;
+    member.prepareUpdate((m) => {
+        m.viewedAt = lastViewed == null ? m.lastViewedAt : lastViewed;
+        m.lastViewedAt = lastViewed == null ? m.lastViewedAt : lastViewed;
+        m.messageCount = messageCount;
+        m.mentionsCount = mentionsCount;
+        m.manuallyUnread = manuallyUnread;
+        m.lastPostAt = lastPostAt;
     });
 
     try {
         if (!prepareRecordsOnly) {
-            const {operator} = DatabaseManager.serverDatabases[serverUrl];
             await operator.batchRecords([member]);
         }
 
-        return member;
+        return {member};
     } catch (error) {
         return {error};
     }
 };
+
+export const updateLastPostAt = async (serverUrl: string, channelId: string, lastPostAt: number, prepareRecordsOnly = false) => {
```

Comment:
I really don't see a use for `lastPostAt` let's chat and see if we can get rid of it